File: 0_1.baekjoon/dp/java/src/BOJ18427.java
import java.io.*;
import java.util.*;

public class BOJ18427 {
    private static final int MOD = 10007;
    private static int N, M, H;
    private static int[][] blocks;
    private static int[][] dp;
    public static void main(String args[]) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        H = Integer.parseInt(st.nextToken());
        dp = new int[N][H + 1];
        blocks = new int[N][];

        for (int i=0;i<N;i++) {
            String[] line = br.readLine().split(" ");
            blocks[i] = new int[line.length];

            for (int j=0;j<line.length;j++) {
                blocks[i][j] = Integer.parseInt(line[j]);
            }
            dp[i][0] = 1;
            for (int j=0;j<=H;j++) {
                dp[i][j] = -1;
            }
        }

        System.out.println(dfs(0, H));
    }

    private static int dfs(int idx, int height) {
        if (height == 0) return 1;
        if (height < 0) return 0;
        if (idx == N) return 0;
        if (dp[idx][height] != -1) return dp[idx][height];


        dp[idx][height] = dfs(idx + 1, height);
        for (int block : blocks[idx]) {
            dp[idx][height] += dfs(idx + 1, height - block);
        }

        dp[idx][height] %= MOD;
        return dp[idx][height];
    }
}
/***
 * dp[h] = 높이 h를 만들 수 있는 블럭의 수
 * blocks[n][m] = n 학생이 가지고 있는 m번째 블럭의 높이
 *
 * dfs(int idx, int h):
 *  if (idx == N || h < 0):
 *      return 0
 * if (dp[h] != 0):
 *      return dp[h]
 *
 * for (int block : blocks[idx]):
 *      dp[h] += dfs(idx + 1, h - block)
 *
 * dp[h] += dfs(idx + 1, h)
 *
 * return dp[h];
 */

File: 0_1.baekjoon/dp/kotlin/src/BOJ18427.kt
private var N = 0;
private var M = 0;
private var H = 0;

private lateinit var dp: Array<IntArray>
private lateinit var blocks: Array<IntArray>
fun main() = with(System.`in`.bufferedReader()) {
    val (n, m, h) = readLine().split(" ").map { it.toInt() }
    N = n;
    M = m;
    H = h;

    dp = Array(N) { IntArray(H + 1) { -1} }
    blocks = Array(N) {
        readLine().split(" ").map { it.toInt() }.toIntArray()
    }

    for (i in 0 until N) {
        dp[i][0] = 1
    }

    print(dfs(0, H))

}

private fun dfs(idx: Int, height: Int): Int {
    if (height == 0) {
        return 1
    }

    if (idx == N || height < 0) {
        return 0
    }

    if (dp[idx][height] != -1) {
        return dp[idx][height]
    }

    dp[idx][height] = dfs(idx + 1, height)
    for (block in blocks[idx]) {
        dp[idx][height] += dfs(idx + 1, height - block)
    }

    dp[idx][height] %= 10007
    return dp[idx][height]
}